keps/sig-instrumentation/0034-distributed-tracing-kep.md

```diff
@@ -0,0 +1,172 @@
+    - [Table of Contents](#table-of-contents)
+    - [Summary](#summary)
+    - [Motivation](#motivation)
+        - [Goals](#goals)
+        - [Non-Goals](#non-goals)
+    - [Proposal](#proposal)
+        - [Architecture](#architecture)
+            - [Trace lifecycle](#trace-lifecycle)
+            - [Root spans](#root-spans)          
+            - [Context propagation](#context-propagation)          
+        - [Out-of-tree changes](#out-of-tree-changes)
+            - [Mutating admission webhook](#mutating-admission-webhook)
+            - [Tracing best-practices documentation](#tracing-best-practices-documentation)
+        - [In-tree changes](#in-tree-changes)
+            - [Trace utility package](#trace-utility-package)
+            - [Basic Kubernetes component trace instrumentation](#basic-kubernetes-component-trace-instrumentation)
+    - [Graduation Requirements](#graduation-requirements)
+    - [Implementation History](#implementation-history)
+
+## Summary
+
+This Kubernetes Enchancement Proposal (KEP) introduces a model for adding distributed tracing to Kubernetes object lifecycles. The inclusion of this trace instrumentation will mark a significant step in making Kubernetes processes more observable, understandable, and debuggable.
+
+
+## Motivation
+
+Debugging latency issues in Kubernetes is an involved process. There are existing tools which can be used to isolate these 
+issues in Kubernetes, but these methods fall short for various reasons. For instance:
+
+* **Logs** are fragmented, and finding out which process was the bottleneck involves digging through troves of unstructured text. In addition, logs do not offer higher-level insight into overall system behavior without an extensive background on the process of interest. 
+* **Events** in Kubernetes are only kept for an hour by default, and don't integrate with visualization of analysis tools. To gain trace-like insights would require a large investment in custom tooling
+* **Latency metrics** are gathered in some places, but these don't provide understanding into _why_ a given process was slow
+
+Distributed tracing, on the other hand, provides a single window into latency information from across many components and plugins. Trace data is structured, and there are numerous established backends for visualizing and querying over it. This KEP would make it possible to, for instance, retrieve and visualize all pod startups that took more than 30 seconds, involved an `nginx` container, and which mounted more than two volumes.
+
+In addition, due to the self-healing nature of Kubernetes, regressions wherein latencies are affected but the overall task is eventually accomplished are not uncommon. With our current monitoring architecture, these "soft regressions" are often difficult to observe and diagnose. Collecting structured trace data on per-object latencies would enable us to detect these long-term regressions automatically, and quickly determine their root causes.
```

Comment:
Can you be more specific in what you mean by "soft regressions" and how the monitoring architecture is not sufficient?

For this specific use case it sounds like to me that a combination of both improving the metrics instrumentation (which is indeed not good enough today) plus sampling "bad" traces would significantly improve the current debugging process.

Reply:
By "soft regression", I mean an issue that doesn't result in a definitive failure, but rather in degraded performance.

You are definitely correct in that even just improving the metrics and sampling bad traces would improve the current process. I think what I was trying to highlight here was that there is potential to plug into existing trace analysis tools to perform automatic root-cause-analysis.  

This could make it possible to, for example, detect a latency regression in pod startup, and then attribute that regression to a change in some metadata (such as a container version, or notice that the regression shows when a pod mounts a certain volume, etc etc). Latency metrics lack the structure / context required to perform this kind of analysis.

Will clarify the KEP on this point.

Reply:
@dashpole on this as well, who might have a better idea on how this will fit in with existing latency metrics

Reply:
The general point here is that in addition to identifying _that_ a regression has occurred, tracing also helps identifies the root-cause of the issue.